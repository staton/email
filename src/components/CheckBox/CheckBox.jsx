import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {MdCheckBoxOutlineBlank, MdCheckBox} from 'react-icons/md';

const propTypes = {
    isChecked: PropTypes.bool,
    onClick: PropTypes.func
};

const defaultProps = {
    isChecked: false,
    onClick: () => {}
};

class CheckBox extends Component {

    constructor() {
        super();

        this.handleClicked = this.handleClicked.bind(this);
    }
	render() {
		return (
            <div 
                className={this.getClassName()}
                onClick={this.handleClicked}>
            {
                (this.props.isChecked) 
                ?   <MdCheckBox /> 
                :   <MdCheckBoxOutlineBlank />
            }
            </div>
		);
    }

    /**
     * Gets the class name(s) for this component.
     * @returns {string} The class name(s).
     */
    getClassName() {
        let className = 'CheckBox';

        if (this.props.isChecked) {
            className += ' CheckBox--checked';
        }

        return className;
    }

    /**
     * Called when the checkbox is clicked.
     * @param {object} e The event.
     */
    handleClicked(e) {
        e.preventDefault();
        e.stopPropagation();
        this.props.onClick();
    }

}

CheckBox.propTypes = propTypes;
CheckBox.defaultProps = defaultProps;

export default CheckBox;