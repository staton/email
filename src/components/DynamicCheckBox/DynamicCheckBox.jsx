import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {MdCheck} from 'react-icons/md';
import BackgroundColor from '../../enums/backgroundColor';

const propTypes = {
    active: PropTypes.bool,
    backgroundColor: PropTypes.string,
    content: PropTypes.string.isRequired,
    isChecked: PropTypes.bool,
    onClick: PropTypes.func
};

const defaultProps = {
    active: false,
    backgroundColor: BackgroundColor.Gray,
    content: null,
    isChecked: false,
    onClick: () => {}
};

class DynamicCheckBox extends Component {

    constructor() {
        super();

        this.handleClicked = this.handleClicked.bind(this);
    }

	render() {
		return (
            <div 
                className={this.getClassName()}
                onClick={this.handleClicked}
                style={{backgroundColor: this.props.backgroundColor}}
            >
            {
                (this.props.active) 
                ?   <span className="DynamicCheckBox__check">
                    {
                        (this.props.isChecked)
                        ?   <MdCheck />
                        :   null
                    }
                    </span>
                :   <span className="DynamicCheckBox__letter">{this.props.content}</span>
            }
            </div>
		);
    }

    /**
     * Gets the class name(s) for this component.
     * @returns {string} The class name(s).
     */
    getClassName() {
        let className = 'DynamicCheckBox';

        if (this.props.isChecked) {
            className += ' DynamicCheckBox--checked';
        }

        if (this.props.active) {
            className += ' DynamicCheckBox--active';
        }
        
        return className;
    }

    /**
     * Called when the checkbox is clicked.
     * @param {object} e The event.
     */
    handleClicked(e) {
        if (this.props.active) {
            e.preventDefault();
            e.stopPropagation();
            this.props.onClick();
        }
    }

}

DynamicCheckBox.propTypes = propTypes;
DynamicCheckBox.defaultProps = defaultProps;

export default DynamicCheckBox;