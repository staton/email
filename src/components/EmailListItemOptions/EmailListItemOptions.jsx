import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';
import _ from 'lodash';
import Email from '../../models/email';
import EmailListItemOptionsButton from '../EmailListItemOptionsButton/EmailListItemOptionsButton';
import {MdDelete, MdFlag, MdMoreHoriz} from 'react-icons/md'

const propTypes = {
    email: PropTypes.instanceOf(Email).isRequired
};

const defaultProps = {
};

export class EmailListItemOptions extends Component {

    constructor() {
        super();

        this.handleDeleteButtonClicked = this.handleDeleteButtonClicked.bind(this);
        this.handleFlagButtonClicked = this.handleFlagButtonClicked.bind(this);
        this.handleMoreButtonClicked = this.handleMoreButtonClicked.bind(this);
    }

	render() {
		return (
            <div className="EmailListItemOptions">
                <EmailListItemOptionsButton 
                    className="EmailListItemOptionsButton__more"
                    content={<MdMoreHoriz />} 
                    onClick={this.handleMoreButtonClicked}
                />
                <EmailListItemOptionsButton 
                    className="EmailListItemOptionsButton__flag"
                    content={<MdFlag />} 
                    onClick={this.handleFlagButtonClicked}
                />
                <EmailListItemOptionsButton 
                    className="EmailListItemOptionsButton__delete"
                    content={<MdDelete />} 
                    onClick={this.handleDeleteButtonClicked}
                />
            </div>
		);
    }

    /**
     * Called when the user taps the Delete button.
     * @param {object} e The event.
     */
    handleDeleteButtonClicked(e) {
        console.log('DELETE clicked');
    }

    /**
     * Called when the user taps the Flag button.
     * @param {object} e The event.
     */
    handleFlagButtonClicked(e) {
        console.log('FLAG clicked');
    }

    /**
     * Called when the user taps the More button.
     * @param {object} e The event.
     */
    handleMoreButtonClicked(e) {
        console.log('MORE clicked');
    }
}

function mapStateToProps(store, ownProps) {
    return {
        email: ownProps.email
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
    },
    dispatch);
}

EmailListItemOptions.propTypes = propTypes;
EmailListItemOptions.defaultProps = defaultProps;

export default connect(mapStateToProps, mapDispatchToProps)(EmailListItemOptions);