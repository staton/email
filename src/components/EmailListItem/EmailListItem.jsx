import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {bindActionCreators} from 'redux';
import {connect} from 'react-redux';
import {MdFlag} from 'react-icons/md';
import CheckBox from '../CheckBox/CheckBox';
import DynamicCheckBox from '../DynamicCheckBox/DynamicCheckBox';
import Email from '../../models/email';
import EmailListItemIcon from '../EmailListItemIcon/EmailListItemIcon';
import EmailListItemOptions from '../EmailListItemOptions/EmailListItemOptions';
import GESTURE_MANAGER from '../../managers/gestureManager';
import {
    selectEmail,
    updateCurrentSwipedEmails
} from '../../redux/actions/email-actions';


const propTypes = {
    email: PropTypes.instanceOf(Email).isRequired,
    isListActive: PropTypes.bool.isRequired,
    isSelected: PropTypes.bool,
    isSwipedOpen: PropTypes.bool,
    setListActive: PropTypes.func.isRequired
};

const defaultProps = {
    isSelected: false,
    isSwipedOpen: false
};

export class EmailListItem extends Component {

    constructor() {
        super();

        this.handleCheckBoxClicked = this.handleCheckBoxClicked.bind(this);
        this.handleLongPressed = this.handleLongPressed.bind(this);
        this.handleTouchEnded = this.handleTouchEnded.bind(this);
        this.handleTouchMoved = this.handleTouchMoved.bind(this);
        this.handleTouchStarted = this.handleTouchStarted.bind(this);
    }

	render() {
        const emailSentDate = this.props.email.EmailSentDateTime.toLocaleDateString();

		return (
            <li 
                className="EmailListItem"
                onTouchEnd={this.handleTouchEnded}
                onTouchMove={this.handleTouchMoved}
                onTouchStart={this.handleTouchStarted}
            >
                <div className={this.getEmailListItemClassNames()}>
                    <div className="EmailListItem__checkbox-container">
                    {this.getCheckBox()}
                    </div>
                    {(!this.props.isSmallScreen) ? this.getFlagIcon() : null}
                    <div className="EmailListItem__email-info-container">
                        {(this.props.isSmallScreen) ? this.getFlagIcon() : null}
                        <div className="EmailListItem__sender-name">{this.props.email.FromName}</div>
                        <div className="EmailListItem__subject">{this.props.email.Subject}</div>
                        <div className="EmailListItem__preview">{this.props.email.Preview}</div>
                        <div className="EmailListItem__sent-date">{emailSentDate}</div>
                    </div>
                </div>
                {(this.props.isSmallScreen) ? <EmailListItemOptions email={this.props.email} /> : null}
            </li>
		);
    }

    /**
     * Gets the checkbox for this list item.
     * @returns {Element} The checkbox.
     */
    getCheckBox() {
        return (
            (this.props.isSmallScreen)
            ?   <DynamicCheckBox 
                    active={this.props.isListActive}
                    backgroundColor={this.props.email.Color}
                    content={this.props.email.FirstLetter}
                    isChecked={this.props.isSelected}
                    onClick={this.handleCheckBoxClicked}
                />
            :   <CheckBox
                    isChecked={this.props.isSelected}
                    onClick={this.handleCheckBoxClicked}
                />
        );
    }

    /**
     * Gets the flag icon for this list item.
     * @returns {Element} The flag icon.
     */
    getFlagIcon() {
        return (
            <EmailListItemIcon 
                content={<MdFlag />} 
                isVisible={this.props.email.Flags.IsImportant}
            />
        );
    }

    /**
     * Gets the class name(s).
     * @returns {String} The class name(s).
     */
    getEmailListItemClassNames() {
        var className = 'EmailListItem__container';
        
        if (!this.props.email.Flags.IsUnread) {
            className += ' EmailListItem__container--read';
        }

        if (this.props.isSwipedOpen) {
            className += ' EmailListItem__container--swiped-open';
        }

        return className;
    }

    /**
     * Gets the flag icon, indicating this is an important email.
     * @returns {Element} The flag icon.
     */
    getFlag() {
        return (
            <EmailListItemIcon 
                content={<MdFlag />} 
                isVisible={this.props.email.Flags.IsImportant}
            />
        );
    }

    /**
     * Called when the user taps the checkbox.
     * @param {object} e The event.
     */
    handleCheckBoxClicked(e) {
        this.props.selectEmail([this.props.email], !this.props.isSelected);
    }

    /**
     * Called when a long press occurs.
     */
    handleLongPressed() {
        if (this.props.currentSwipedEmails.length > 0) {
            // automatically reset any currently swiped emails:
            this.props.updateCurrentSwipedEmails(this.props.currentSwipedEmails, false);
        }
        
        this.props.setListActive(true);
        this.props.selectEmail([this.props.email], true);
    }

    /**
     * Called when a touch action ends.
     * @param {object} e The event.
     */
    handleTouchEnded(e) {
        if (this.props.isListActive)
            return;
            
        GESTURE_MANAGER.handleTouchEnded(e);
        GESTURE_MANAGER.checkForSwipe(
            () => this.props.updateCurrentSwipedEmails([this.props.email], true),
            () => this.props.updateCurrentSwipedEmails([this.props.email], false)
        );
    }

    /**
     * Called when a touch action begins.
     * @param {object} e The event.
     */
    handleTouchMoved(e) {
        if (this.props.isListActive)
            return;

        GESTURE_MANAGER.handleTouchMoved(e);
    }

    /**
     * Called when a touch action begins.
     * @param {object} e The event.
     */
    handleTouchStarted(e) {
        if (this.props.isListActive)
            return;

        let id = GESTURE_MANAGER.handleTouchStarted(e);
        GESTURE_MANAGER.startLongPressTimer(id, this.handleLongPressed);
    }

}

function mapStateToProps(store, ownProps) {
    return {
        currentSwipedEmails: store.email.currentSwipedEmails,
        email: ownProps.email,
        isListActive: ownProps.isListActive,
        isSelected: ownProps.isSelected,
        isSmallScreen: store.app.isSmallScreen,
        isSwipedOpen: ownProps.isSwipedOpen,
        selectEmail: store.email.selectEmail,
        setListActive: ownProps.setListActive
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
        selectEmail: selectEmail,
        updateCurrentSwipedEmails: updateCurrentSwipedEmails
    },
    dispatch);
}

EmailListItem.propTypes = propTypes;
EmailListItem.defaultProps = defaultProps;

export default connect(mapStateToProps, mapDispatchToProps)(EmailListItem);