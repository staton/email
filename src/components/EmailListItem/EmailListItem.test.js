import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';
import { EmailListItem } from './EmailListItem';
import Email from '../../models/email';
import EmailFlags from '../../models/emailFlags';
import CheckBox from '../CheckBox/CheckBox';
import DynamicCheckBox from '../DynamicCheckBox/DynamicCheckBox';
import EmailListItemOptions from '../EmailListItemOptions/EmailListItemOptions';

describe('<EmailListItem />', () => {

    const GET_EMAIL = () => new Email(
        0, 'bob', 'bob@example.com', ['s@gmail.com'], [], [], 'hello', 'preview', 
        new Date('2018-05-05T12:05:00Z'), null, new EmailFlags(false, false, false, false));

    it('renders correctly', () => {
        const wrapper = shallow(
            <EmailListItem 
                email={GET_EMAIL()}
                isListActive={false}
                setListActive={jest.fn}
            />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('contains a CheckBox for large screen devices', () => {
        const wrapper = shallow(
            <EmailListItem 
                email={GET_EMAIL()}
                isListActive={false}
                isSmallScreen={false}
                setListActive={jest.fn}
            />);
        expect(wrapper.find(CheckBox).length).toEqual(1);
        expect(wrapper.find(DynamicCheckBox).length).toEqual(0);
    });

    it('contains a DynamicCheckBox for small screen devices', () => {
        const wrapper = shallow(
            <EmailListItem 
                email={GET_EMAIL()}
                isListActive={false}
                isSmallScreen={true}
                setListActive={jest.fn}
            />);
        expect(wrapper.find(DynamicCheckBox).length).toEqual(1);
        expect(wrapper.find(CheckBox).length).toEqual(0);
    });

    it.skip('handles touch and click events', () => {
        const onClick = jest.fn();
        const onTouchEnd = jest.fn();
        const onTouchMove = jest.fn();
        const onTouchStart = jest.fn();

        const wrapper = mount(
            <EmailListItem 
                email={GET_EMAIL()}
                isListActive={false}
                isSmallScreen={true}
                setListActive={jest.fn}
                onClick={onClick}
                onTouchEnd={onTouchEnd}
                onTouchMove={onTouchMove}
                onTouchStart={onTouchStart}
            />);

        wrapper.simulate('click');
        expect(onClick).toHaveBeenCalledTimes(1);
        wrapper.simulate('touchend');
        expect(onTouchEnd).toHaveBeenCalledTimes(1);
        wrapper.simulate('touchmove');
        expect(onTouchMove).toHaveBeenCalledTimes(1);
        wrapper.simulate('touchstart');
        expect(onTouchStart).toHaveBeenCalledTimes(1);
    });

    it('has the ..--read class name if the email is read', () => {
        const email = GET_EMAIL();
        email.Flags.IsUnread = false;
        const wrapper = shallow(
            <EmailListItem 
                email={email}
                isListActive={false}
                setListActive={jest.fn}
            />);
        expect(wrapper.find('.EmailListItem__container--read').length).toEqual(1);
    });

    it('does not have the ..--read class name if the email is unread', () => {
        const email = GET_EMAIL();
        email.Flags.IsUnread = true;
        const wrapper = shallow(
            <EmailListItem 
                email={email}
                isListActive={false}
                setListActive={jest.fn}
            />);
        expect(wrapper.find('.EmailListItem__container--read').length).toEqual(0);
    });

    it('has the ..--swiped-open class name if the email is swiped open', () => {
        const wrapper = shallow(
            <EmailListItem 
                email={GET_EMAIL()}
                isListActive={false}
                isSwipedOpen={true}
                setListActive={jest.fn}
            />);
        expect(wrapper.find('.EmailListItem__container--swiped-open').length).toEqual(1);
    });

    it('does not have the ..--swiped-open class name if the email is not swiped open', () => {
        const wrapper = shallow(
            <EmailListItem 
                email={GET_EMAIL()}
                isListActive={false}
                isSwipedOpen={false}
                setListActive={jest.fn}
            />);
        expect(wrapper.find('.EmailListItem__container--swiped-open').length).toEqual(0);
    });

    it('has EmailListItemOptions for small screen devices', () => {
        const wrapper = shallow(
            <EmailListItem 
                email={GET_EMAIL()}
                isListActive={false}
                isSmallScreen={true}
                setListActive={jest.fn}
            />);
        expect(wrapper.find(EmailListItemOptions).length).toEqual(1);
    });

    it('does not have EmailListItemOptions for large screen devices', () => {
        const wrapper = shallow(
            <EmailListItem 
                email={GET_EMAIL()}
                isListActive={false}
                isSmallScreen={false}
                setListActive={jest.fn}
            />);
        expect(wrapper.find(EmailListItemOptions).length).toEqual(0);
    });

});