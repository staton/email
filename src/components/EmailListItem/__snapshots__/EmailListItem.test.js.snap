// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailListItem /> renders correctly 1`] = `
<li
  className="EmailListItem"
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="EmailListItem__container EmailListItem__container--read"
  >
    <div
      className="EmailListItem__checkbox-container"
    >
      <CheckBox
        isChecked={false}
        onClick={[Function]}
      />
    </div>
    <EmailListItemIcon
      content={<Unknown />}
      isVisible={false}
    />
    <div
      className="EmailListItem__email-info-container"
    >
      <div
        className="EmailListItem__sender-name"
      >
        bob
      </div>
      <div
        className="EmailListItem__subject"
      >
        hello
      </div>
      <div
        className="EmailListItem__preview"
      >
        preview
      </div>
      <div
        className="EmailListItem__sent-date"
      >
        2018-5-5
      </div>
    </div>
  </div>
</li>
`;
